File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    let answer = n - 1;
    const queue = [1];
    const graph = Array.from({length: n+1}, () => []);
    
    edge.forEach(([a, b]) => {
        graph[a].push(b);
        graph[b].push(a);
    });
    
    while (true) {
        [...queue].forEach(node => {
            queue.shift();
            if (graph[node].length) {
                queue.push(...graph[node]);
                graph[node] = [];
                n--;
            }
        });
        
        if (n === 0) return answer;
        answer = n;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = n - 1
    queue = deque([1])
    graph = [[] for _ in range(n+1)]
    
    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)
    
    while True:
        for _ in range(len(queue)):
            node = queue.popleft()
            if graph[node]:
                queue.extend(graph[node])
                graph[node] = []
                n -= 1
        
        if n == 0: return answer
        answer = n